core/circuit.go

```diff
@@ -0,0 +1,262 @@
+			cc.Log.TRACE.Printf("creating circuit from circuitRef: %s", cc.Circuits[ccId].Name)
+			cc.Circuits[ccId].parentCircuit = cc
+			if err := cc.Circuits[ccId].InitCircuits(site, cp); err != nil {
+				return err
+			}
+			if vmtr := cc.GetVMeter(); vmtr != nil {
+				vmtr.AddConsumer(&cc.Circuits[ccId])
+			}
+			cc.Circuits[ccId].PrintCircuits(0)
+		}
+	} else {
+		cc.Log.TRACE.Printf("no sub circuits")
+	}
+	cc.Log.TRACE.Println("InitCircuits() exit")
+	cc.Log.INFO.Printf("initialized new circuit: %s, limit: %.1fA", cc.Name, cc.MaxCurrent)
+	return nil
+}
+
+func (cc *Circuit) GetVMeter() *VMeter {
+	return cc.vMeter
+}
+
+// trace output of circuit and subcircuits
+func (cc *Circuit) PrintCircuits(indent int) {
+	for _, s := range cc.DumpConfig(0, 15) {
+		cc.Log.TRACE.Println(s)
+	}
+}
+
+// returns string array to dump the config
+func (cc *Circuit) DumpConfig(indent int, maxIndent int) []string {
+
+	var res []string
+
+	cfgDump := fmt.Sprintf("%s%s:%s meter %s maxCurrent %.1fA",
+		strings.Repeat(" ", indent),
+		cc.Name,
+		strings.Repeat(" ", maxIndent-len(cc.Name)-indent),
+		presence[cc.GetVMeter() == nil],
+		cc.MaxCurrent,
+	)
+	res = append(res, cfgDump)
+
+	// cc.Log.TRACE.Printf("%s%s%s: (%p) log: %t, meter: %t, parent: %p\n", strings.Repeat(" ", indent), cc.Name, strings.Repeat(" ", 10-indent), cc, cc.Log != nil, cc.meterCurrent != nil, cc.parentCircuit)
+	for id := range cc.Circuits {
+		// this does not work (compiler error), but linter requests it. Github wont build ...
+		// res = append(res, cc.Circuits[id].DumpConfig(indent+2, maxIndent))
+		// hacky work around
+		for _, l := range cc.Circuits[id].DumpConfig(indent+2, maxIndent) {
+			res = append(res, l)
+			// add useless command
+			time.Sleep(0)
+		}
+	}
+	return res
+}
+
+// returns the circiut with given name, checking all subcircuits
+func (cc *Circuit) GetCircuit(n string) *Circuit {
+	cc.Log.TRACE.Printf("searching for circuit %s in %s", n, cc.Name)
+	if cc.Name == n {
+		cc.Log.TRACE.Printf("found circuit %s (%p)", cc.Name, &cc)
+		return cc
+	} else {
+		for ccId := range cc.Circuits {
+			cc.Log.TRACE.Printf("start looking in circuit %s (%p)", cc.Circuits[ccId].Name, &cc.Circuits[ccId])
+			retCC := cc.Circuits[ccId].GetCircuit(n)
+			if retCC != nil {
+				cc.Log.TRACE.Printf("found circuit %s (%p)", retCC.Name, &retCC)
+				return retCC
+			}
+		}
+	}
+	cc.Log.INFO.Printf("could not find circuit %s", n)
+	return nil
+}
+
+// publish sends values to UI and databases
+func (cc *Circuit) publish(key string, val interface{}) {
+	// test helper
+	if cc.uiChan == nil {
+		return
+	}
+
+	cc.uiChan <- util.Param{
+		Circuit: &cc.Name,
+		Key:     key,
+		Val:     val,
+	}
+}
+
+// set the UI channel to publish information
+func (cc *Circuit) Prepare(uiChan chan<- util.Param) {
+	cc.uiChan = uiChan
+	cc.publish("name", cc.Name)
+	cc.publish("maxCurrent", cc.MaxCurrent)
+	if vmtr := cc.GetVMeter(); vmtr != nil {
+		cc.publish("virtualMeter", true)
```

Comment:
brauchts das im UI?

Reply:
ist fürs status. Ich kann auch in der Konfig nachlesen. Brauchts Namen im status? Machts etwas einfacher ... up 2u